pkg/rbd/rbd_util.go

```diff
@@ -205,9 +224,430 @@ func rbdStatus(pOpts *rbdVolume, userID string, credentials map[string]string) (
+	err = util.RemoveOMapKey(rbdVol.Monitors, rbdVol.AdminID, key, rbdVol.Pool,
+		volOMap, volOMapNameKey+rbdVol.RequestName)
+	if err != nil {
+		klog.V(4).Infof("failed removing oMap key %s (%s)", volOMapNameKey+rbdVol.RequestName, err)
+		return err
+	}
+
+	return nil
+}
+
+/*
+reserverRBDSnap and reserveRBDVol add respective entries to the volumes and snapshots omaps, post
+generating a target snapshot or image name for use. Further, these functions create the snapshot or
+image name omaps, to store back pointers to the CSI generated request names.
+
+This is performed within the request name lock, to ensure that requests with the same name do not
+manipulate the omap entries concurrently.
+*/
+func reserveRBDSnap(rbdSnap *rbdSnapshot, credentials map[string]string) error {
+	var vi util.VolumeIdentifier
+
+	// generate a uuid for the snap image name
+	uuid := uuid.NewUUID().String()
+	snapName := snapNameHeader + uuid
+
+	key, err := getRBDKey(rbdSnap.ClusterID, rbdSnap.AdminID, credentials)
+	if err != nil {
+		return err
+	}
+
+	poolID, err := util.GetPoolID(rbdSnap.Monitors, rbdSnap.AdminID, key,
+		rbdSnap.Pool)
+	if err != nil {
+		klog.V(4).Infof("Error fetching pool ID (%s)", err)
+		return err
+	}
+
+	// Create request snapName key in csi snaps omap and store the uuid based
+	// snap name into it
+	err = util.SetOMapKeyValue(rbdSnap.Monitors, rbdSnap.AdminID, key,
+		rbdSnap.Pool, snapsOMap, snapsOMapNameKey+rbdSnap.RequestName, snapName)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err != nil {
+			klog.Warningf("reservation failed for volume: %s", rbdSnap.RequestName)
+			errDefer := unreserveRBDSnap(rbdSnap, credentials)
+			if errDefer != nil {
+				klog.Warningf("failed undoing reservation of snapshot: %s", rbdSnap.RequestName)
+			}
+		}
+	}()
+
+	// Create snap name based omap and store CSI request name key and source information
+	err = util.SetOMapKeyValue(rbdSnap.Monitors, rbdSnap.AdminID, key, rbdSnap.Pool,
+		snapImgOMap+snapName, snapImgOMapNameKey, rbdSnap.RequestName)
+	if err != nil {
+		return err
+	}
+	err = util.SetOMapKeyValue(rbdSnap.Monitors, rbdSnap.AdminID, key, rbdSnap.Pool,
+		snapImgOMap+snapName, snapImgOMapParentKey, rbdSnap.VolName)
+	if err != nil {
+		return err
+	}
+
+	// generate the volume ID to return to the CO system
+	vi = util.VolumeIdentifier{
+		PoolID:          poolID,
+		EncodingVersion: volIDVersion,
+		ClusterID:       rbdSnap.ClusterID,
+		ImageName:       snapName,
+	}
+	rbdSnap.SnapID, err = vi.ComposeVolID()
+	if err != nil {
+		return err
+	}
+	rbdSnap.SnapName = snapName
+	klog.V(4).Infof("Generated Volume ID (%s) and image name (%s) for request name (%s)",
+		rbdSnap.SnapID, rbdSnap.VolName, rbdSnap.RequestName)
+
+	return nil
+}
+
+func reserveRBDVol(rbdVol *rbdVolume, credentials map[string]string) error {
+	var vi util.VolumeIdentifier
+
+	// generate a uuid for the image name
+	uuid := uuid.NewUUID().String()
```

Comment:
Nit: same comment here re: package name

Reply:
Will address in a future update to the PR.